packages/jest-react/src/JestReact.js

```diff
@@ -0,0 +1,142 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import jestDiff from 'jest-diff';
+
+import {REACT_ELEMENT_TYPE, REACT_FRAGMENT_TYPE} from 'shared/ReactSymbols';
+
+function captureAssertion(fn) {
+  // Trick to use a Jest matcher inside another Jest matcher. `fn` contains an
+  // assertion; if it throws, we capture the error and return it, so the stack
+  // trace presented to the user points to the original assertion in the
+  // test file.
+  try {
+    fn();
+  } catch (error) {
+    return {
+      pass: false,
+      message: () => error.message,
+    };
+  }
+  return {pass: true};
+}
+
+export function toFlushAll(renderer, expectedYields) {
+  const actualYields = renderer.unstable_flushAll();
+  return captureAssertion(() => expect(actualYields).toEqual(expectedYields));
+}
+
+export function toFlushThrough(renderer, expectedYields) {
+  const actualYields = renderer.unstable_flushNumberOfYields(
+    expectedYields.length,
+  );
+  return captureAssertion(() => expect(actualYields).toEqual(expectedYields));
+}
+
+export function toClearYields(ReactTestRenderer, expectedYields) {
+  const actualYields = ReactTestRenderer.unstable_clearYields();
+  return captureAssertion(() => expect(actualYields).toEqual(expectedYields));
+}
+
+export function toFlushAndThrow(renderer, expectedYields, ...rest) {
+  return captureAssertion(() => {
+    try {
+      expect(() => {
+        renderer.unstable_flushAll();
+      }).toThrow(...rest);
+    } catch (error) {
+      const actualYields = renderer.unstable_clearYields();
+      expect(actualYields).toEqual(expectedYields);
+      throw error;
+    }
+    const actualYields = renderer.unstable_clearYields();
+    expect(actualYields).toEqual(expectedYields);
+  });
+}
+
+export function toMatchRenderedOutput(renderer, expectedJSX) {
+  const actualJSON = renderer.toJSON();
+
+  let actualJSX;
```

Comment:
The tricky bit is that ReactTestRenderer.create will run the renderer synchronously on a new root, so it will cause React to interrupt the previous root and switch to a new one. That makes it unsuitable for some tests, and it’s a non-obvious/surprising behavior when it happens.

Reply:
That's fair. I think it's reasonable

I'm concerned that what we have with this PR isn't flexible enough to be very useful with many "real" components.

Reply:
Oh whoops, missed your second comment.

> Also why wouldn't that test work?

Lots of undefined property access errors. If you think it should be _possible_ for it to work with more complex test cases, then I can help dig into how we'd fix it.

Reply:
Yeah it's probably just a bug. I'll add more tests as I go.

Reply:
Cool. In that case, it would be great to update `ErrorBoundaryReconciliation` to use the new matcher.

Reply:
Yeah it was just a bug :)

Reply:
Nice!